node/impl/full/eth.go

```diff
@@ -0,0 +1,828 @@
+	to, err := ethAddr.ToFilecoinAddress()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot get Filecoin address: %w", err)
+	}
+
+	// use the system actor as the caller
+	from, err := address.NewIDAddress(0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct system sender address: %w", err)
+	}
+	msg := &types.Message{
+		From:       from,
+		To:         to,
+		Value:      big.Zero(),
+		Method:     builtintypes.MethodsEVM.GetBytecode,
+		Params:     nil,
+		GasLimit:   build.BlockGasLimit,
+		GasFeeCap:  big.Zero(),
+		GasPremium: big.Zero(),
+	}
+
+	ts := a.Chain.GetHeaviestTipSet()
+
+	// Try calling until we find a height with no migration.
+	var res *api.InvocResult
+	for {
+		res, err = a.StateManager.Call(ctx, msg, ts)
+		if err != stmgr.ErrExpensiveFork {
+			break
+		}
+		ts, err = a.Chain.GetTipSetFromKey(ctx, ts.Parents())
+		if err != nil {
+			return nil, xerrors.Errorf("getting parent tipset: %w", err)
+		}
+	}
+
+	if err != nil {
+		// if the call resulted in error, this is not an EVM smart contract;
+		// return no bytecode.
+		return nil, nil
+	}
+
+	if res.MsgRct == nil {
+		return nil, fmt.Errorf("no message receipt")
+	}
+
+	if res.MsgRct.ExitCode.IsError() {
+		return nil, xerrors.Errorf("message execution failed: exit %s, reason: %s", res.MsgRct.ExitCode, res.Error)
+	}
+
+	var bytecodeCid cbg.CborCid
+	if err := bytecodeCid.UnmarshalCBOR(bytes.NewReader(res.MsgRct.Return)); err != nil {
+		return nil, fmt.Errorf("failed to decode EVM bytecode CID: %w", err)
+	}
+
+	blk, err := a.Chain.StateBlockstore().Get(ctx, cid.Cid(bytecodeCid))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get EVM bytecode: %w", err)
+	}
+
+	return blk.RawData(), nil
+}
+
+func (a *EthModule) EthGetStorageAt(ctx context.Context, ethAddr api.EthAddress, position api.EthBytes, blkParam string) (api.EthBytes, error) {
+	l := len(position)
+	if l > 32 {
+		return nil, fmt.Errorf("supplied storage key is too long")
+	}
+
+	// pad with zero bytes if smaller than 32 bytes
+	position = append(make([]byte, 32-l, 32-l), position...)
+
+	to, err := ethAddr.ToFilecoinAddress()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot get Filecoin address: %w", err)
+	}
+
+	// use the system actor as the caller
+	from, err := address.NewIDAddress(0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct system sender address: %w", err)
+	}
+
+	// TODO super duper hack (raulk). The EVM runtime actor uses the U256 parameter type in
+	//  GetStorageAtParams, which serializes as a hex-encoded string. It should serialize
+	//  as bytes. We didn't get to fix in time for Iron, so for now we just pass
+	//  through the hex-encoded value passed through the Eth JSON-RPC API, by remarshalling it.
+	//  We don't fix this at origin (builtin-actors) because we are not updating the bundle
+	//  for Iron.
+	tmp, err := position.MarshalJSON()
```

Comment:
Ideally we'd fix https://github.com/filecoin-project/ref-fvm/issues/912.

Reply:
Done.